### PR TITLE
Add Atmel SAM I2S (Inter-IC Sound) driver based on SSC module

### PR DESCRIPTION
**Pull request by mnkp**
This PR comes in three parts: DMA driver used by I2S, I2S driver itself and I2S testcases. The testcases are not yet complete however bulk of the functionality is already tested.

### CODE REVIEW

**Comment by mnkp** [comment]
Hi @galak, it would be great if you could have a look at this PR. I don't want it to be forgotten. If it's of any help I can submit Atmel SAM DMA driver in a separate PR. All together it's certainly a large chunk of code.

File: drivers/dma/xdmac_sam.c
Comment (tbursztyka) [inline drivers/dma/xdmac_sam.c]: add {} even on one-liner statement for if/while/for. Apply that everywhere.

That said, it's weird that checkpatch did not catch those and make the checks failing...

File: drivers/dma/xdmac_sam.c
Comment (tbursztyka) [inline drivers/dma/xdmac_sam.c]: irq_enable() is usually called in relevant irq_config, with a Kconfig (or soc.h) based IRQ number (which saves a bit of ROM in device's data structure). Unless you've got a good reason to do otherwise.

File: drivers/dma/xdmac_sam.c
Comment (tbursztyka) [inline drivers/dma/xdmac_sam.c]: for instance here, irq_enable(XDMAC_IRQn); would be just fine.

File: drivers/i2s/i2s_sam_ssc.c
Comment (tbursztyka) [inline drivers/i2s/i2s_sam_ssc.c]: same comment as in xdmac_sam.c

File: drivers/dma/xdmac_sam.c
Comment (tbursztyka) [inline drivers/dma/xdmac_sam.c]: Rename your driver as dma_xdmac_sam.c

Not sure it's written anywhere but in drivers, we always write names as <domain>_<name>.<c/h>

**Review by tbursztyka — CHANGES_REQUESTED** [review]
Tiny comments from me.

@mnkp : add relevant reviewers for your PR, or then it might hang there indefinitely.

After a quick glance, besides the comments I had, this PR sounds fairly ok to me. But getting reviews from someone who works around DMA or I²S would be welcome.

File: drivers/dma/xdmac_sam.c
Comment (mnkp) [reply drivers/dma/xdmac_sam.c]: I'll change the name of the driver to dma_sam_xdmac.c to match the other Atmel SAM drivers. That's the second of the two conventions we're using.

File: drivers/dma/xdmac_sam.c
Comment (mnkp) [reply drivers/dma/xdmac_sam.c]: Sorry, I overlooked this one.

**Review by mnkp — COMMENTED** [review]
@tbursztyka thanks for the review. Unfortunately I don't have a permission to add reviewers to this PR. Shouldn't it be done automatically by the script based on the CODEOWNERS file?

**Comment by tbursztyka** [comment]
> Shouldn't it be done automatically by the script based on the CODEOWNERS file?

You're right. Which means the script has a bug then because: 

`drivers/*/*qmsi*                         @nashif`

And there is a qmsi driver in drivers/dma so @nashif  should have been added.

File: drivers/dma/xdmac_sam.c
Comment (mnkp) [reply drivers/dma/xdmac_sam.c]: I placed `irq_enable()` inside initialization function to ensure that `IRQ_CONNECT` is the only item called by dma0_sam_irq_config. I heard several times already that we should re-implement IRQ_CONNECT macro in a way that it does not have to be called within a function. In this case we could get rid of dma0_sam_irq_config function and `dev_cfg->irq_config()` call.

If this is ever done updating the driver will be much easier if we keep the current arrangement. But maybe I'm going to far? Most other drivers do call `irq_enable()` within IRQ configuration function.

File: tests/drivers/i2s/i2s_api/src/main.c
Comment (mnkp) [inline tests/drivers/i2s/i2s_api/src/main.c]: I placed function declarations here following an example of other tests. However, that does not seem very handy. Maybe we should place all these function declarations in a separate header file? Or define `ztest_test_suite` in a testcase .c file? E.g. place `ztest_test_suite(i2s_loopback_test, ...);` in test_i2s_loopback.c so there is no need to write test function declarations.

File: drivers/dma/xdmac_sam.c
Comment (andrewboie) [reply drivers/dma/xdmac_sam.c]: >  I heard several times already that we should re-implement IRQ_CONNECT macro in a way that it does not have to be called within a function

Heard from who, exactly?
IRQ_CONNECT() has a runtime component. You can't pull it out of a function body. It does more that just declare data.

File: tests/drivers/i2s/i2s_speed/src/test_i2s_speed.c
Comment (mnkp) [inline tests/drivers/i2s/i2s_speed/src/test_i2s_speed.c]: Maximum interface speed should be a parameter which is defined independently for every SPI driver / board and not hard coded in a .c file. I did not found however a good example of how to define such testcase parameters. Shall I use some Kconfig test option, e.g. TEST_I2S_FRAME_CLK_FREQ_MAX? If yes, where should I set its value?

File: drivers/dma/xdmac_sam.c
Comment (mnkp) [reply drivers/dma/xdmac_sam.c]: Last time it was @andyross who stated in PR #1426
> Honestly IRQ_CONNECT() needs some rethinking as a framework.

It's not such a big issue but if we did not have to define a per driver instance IRQ configuration function our life would be slightly easier.

Comment (mnkp) [reply drivers/dma/xdmac_sam.c]: > irq_enable() is usually called in relevant irq_config, with a Kconfig (or soc.h) based IRQ number (which saves a bit of ROM in device's data structure)

I run rom_report. At present dma_sam_xdmac.c takes 670 bytes of ROM. After placing irq_enable() in dma0_sam_irq_config() it took 676 bytes of ROM. So the current code actually generates a smaller image, probably because XDMAC_IRQn is stored here as u8_t next to periph_id which is also u8_t. Otherwise this two values would each occupy 4 bytes.

I'll leave this code as it is.

**Comment by mnkp** [comment]
I pushed new version fixing issues reported by @tbursztyka and rebased on latest 'arm' branch.

File: arch/arm/soc/atmel_sam/same70/soc.h
Comment (galak) [inline arch/arm/soc/atmel_sam/same70/soc.h]: are these id's unique from peripheral-id we have in the dts for each device node?

**Review by galak — CHANGES_REQUESTED** [review]
Can we get device tree support for the dma controller (at least just a node for the controller itself).  We will need to work up channel support in the future.

**Comment by mnkp** [comment]
@galak yes, let's aim for 1.10 release. This PR was already way too long in the waiting. I'll prepare an updated version with DTS support for DMA controller in the next few days.

**Comment by jfischer-phytec-iot** [comment]
@mnkp I started with DTS support for DMA, currently I do not have time for it, you can take it over: https://github.com/jfischer-phytec-iot/zephyr/tree/wip-dma-dts ping me (IRC) if you have any questions

**Comment by nashif** [comment]
@mnkp can you please port this to cmake?

**Comment by mnkp** [comment]
I've ported this PR to cmake. I'll need some more time to add DTS support for DMA controller. Since  DTS support for DMA controller is completely missing in Zephyr we could merge this PR as it is and add  DTS support later.

File: tests/drivers/i2s/i2s_api/src/Makefile
Comment (nashif) [inline tests/drivers/i2s/i2s_api/src/Makefile]: remove this makefile

File: tests/drivers/i2s/i2s_speed/src/Makefile
Comment (nashif) [inline tests/drivers/i2s/i2s_speed/src/Makefile]: remove this file

File: tests/drivers/i2s/i2s_speed/testcase.yaml
Comment (nashif) [inline tests/drivers/i2s/i2s_speed/testcase.yaml]: can you please the indentation here, 2 space for yaml

**Review by nashif — CHANGES_REQUESTED** [review]

File: tests/drivers/i2s/i2s_api/src/Makefile
Comment (mnkp) [reply tests/drivers/i2s/i2s_api/src/Makefile]: This looks like a github bug, I did remove this file. My commit doesn't have it.

Comment (mnkp) [reply tests/drivers/i2s/i2s_api/src/Makefile]: Actually my commit does have this Makefile, even if it's missing from workarea and git status doesn't report it. Sorry about that.

File: tests/drivers/i2s/i2s_speed/testcase.yaml
Comment (mnkp) [reply tests/drivers/i2s/i2s_speed/testcase.yaml]: Looking at other `testcase.yaml` files, e.g. from i2c test I understand that the problem was with
```
  - test:
```
line. This is fixed now.

**Review by nashif — APPROVED** [review]

**Comment by galak** [comment]
This got merged into the arm branch and don't have DTS support.  I'm willing to let them in with the understanding that we will fixup the DTS support in the future.

**Pull request merged**
